File: android/src/main/java/com/bluebillywig/bbplayer/BBShortsView.kt
package com.bluebillywig.bbplayer

import android.app.Activity
import android.util.Log
import android.widget.FrameLayout
import com.bluebillywig.bbnativeplayersdk.BBNativeShorts
import com.bluebillywig.bbnativeplayersdk.BBNativeShortsView
import com.bluebillywig.bbnativeplayersdk.BBNativeShortsViewDelegate
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.uimanager.events.Event

// Debug logging helper - only logs in debug builds
private inline fun debugLog(tag: String, message: () -> String) {
    if (BuildConfig.DEBUG) {
        Log.d(tag, message())
    }
}

/**
 * React Native View for Blue Billywig Native Shorts
 *
 * This is a separate component from BBPlayerView specifically for Shorts playback.
 * Shorts use a different native SDK view (BBNativeShortsView) that supports
 * vertical swipe navigation and TikTok-style full-screen video experience.
 *
 * Usage:
 * ```tsx
 * <BBShortsView
 *   jsonUrl="https://demo.bbvms.com/sh/58.json"
 *   onDidSetupWithJsonUrl={(url) => console.log('Shorts loaded:', url)}
 *   onDidFailWithError={(error) => console.error('Error:', error)}
 * />
 * ```
 */
class BBShortsView(private val reactContext: ThemedReactContext) : FrameLayout(reactContext),
    BBNativeShortsViewDelegate {

    private val currentActivity: Activity
        get() = reactContext.currentActivity ?: throw IllegalStateException("No current activity")

    private var jsonUrl: String = ""
    private var options = HashMap<String, Any?>()
    private var shortsSetup: Boolean = false

    private var shortsView: BBNativeShortsView? = null

    init {
        // Default to black background for Shorts
        setBackgroundColor(android.graphics.Color.BLACK)
    }

    /**
     * Override onLayout to ensure child views receive proper bounds.
     * This is critical for React Native Fabric where layout isn't automatically propagated.
     */
    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        // Explicitly layout all children to fill the container
        for (i in 0 until childCount) {
            getChildAt(i)?.layout(0, 0, right - left, bottom - top)
        }
    }

    // Event emission helper using modern EventDispatcher
    private fun sendEvent(eventName: String, params: WritableMap?) {
        val surfaceId = UIManagerHelper.getSurfaceId(reactContext)
        val eventDispatcher = UIManagerHelper.getEventDispatcherForReactTag(reactContext, id)
        eventDispatcher?.dispatchEvent(BBShortsEvent(surfaceId, id, eventName, params))
    }

    private fun sendEvent(eventName: String) {
        sendEvent(eventName, null)
    }

    /**
     * Custom event class for BBShorts events.
     * Uses the modern Event API instead of deprecated RCTEventEmitter.
     */
    private class BBShortsEvent(
        surfaceId: Int,
        viewTag: Int,
        private val name: String,
        private val data: WritableMap?
    ) : Event<BBShortsEvent>(surfaceId, viewTag) {

        override fun getEventName(): String = name

        override fun getEventData(): WritableMap = data ?: Arguments.createMap()
    }

    fun setJsonUrl(url: String) {
        val oldUrl = this.jsonUrl
        Log.d("BBShortsView", "setJsonUrl: $url (current: $oldUrl)")

        // If URL changed and we already have a shorts view, recreate it
        if (oldUrl != url && shortsSetup) {
            Log.d("BBShortsView", "jsonUrl changed - recreating shorts view")
            this.jsonUrl = url
            removeShorts()
            createAndAddShorts()
        } else {
            this.jsonUrl = url
        }
    }

    fun setOptions(optionsMap: ReadableMap?) {
        optionsMap?.let { map ->
            val iterator = map.keySetIterator()
            while (iterator.hasNextKey()) {
                val key = iterator.nextKey()
                when (map.getType(key)) {
                    com.facebook.react.bridge.ReadableType.Boolean -> options[key] = map.getBoolean(key)
                    com.facebook.react.bridge.ReadableType.Number -> options[key] = map.getDouble(key)
                    com.facebook.react.bridge.ReadableType.String -> options[key] = map.getString(key)
                    else -> {}
                }
            }
        }
    }

    fun setupShorts() {
        Log.d("BBShortsView", "setupShorts called - jsonUrl: $jsonUrl, shortsSetup: $shortsSetup")

        if (jsonUrl.isEmpty()) {
            Log.w("BBShortsView", "Cannot setup shorts - jsonUrl is empty")
            return
        }

        if (!shortsSetup) {
            // Initial setup - create the shorts view with the initial URL
            Log.d("BBShortsView", "Initial setup - creating shorts view")
            createAndAddShorts()
        }
    }

    private fun createAndAddShorts() {
        debugLog("BBShortsView") { "Creating BBNativeShortsView with factory method" }

        // Convert options to HashMap for the native SDK
        val nativeOptions: HashMap<String, Any?>? = if (options.isNotEmpty()) options else null

        shortsView = BBNativeShorts.createShortsView(currentActivity, jsonUrl, nativeOptions)
        shortsView?.delegate = this@BBShortsView

        addView(shortsView, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))

        shortsSetup = true
        debugLog("BBShortsView") { "Shorts setup complete with URL: $jsonUrl" }
    }

    private fun removeShorts() {
        debugLog("BBShortsView") { "removeShorts called" }
        shortsSetup = false
        removeAllViews()

        shortsView?.let {
            it.__destruct()
        }
        shortsView = null
    }

    fun destroy() {
        Log.d("BBShortsView", "destroy() called")
        removeShorts()
    }

    // BBNativeShortsViewDelegate implementations
    override fun didSetupWithJsonUrl(shortsView: BBNativeShortsView, url: String?) {
        debugLog("BBShortsView") { "didSetupWithJsonUrl: $url" }

        val params = Arguments.createMap().apply {
            putString("url", url)
        }
        sendEvent("onDidSetupWithJsonUrl", params)
    }

    override fun didFailWithError(shortsView: BBNativeShortsView, error: String?) {
        Log.e("BBShortsView", "didFailWithError: $error")

        val params = Arguments.createMap().apply {
            putString("error", error ?: "Unknown error occurred")
        }
        sendEvent("onDidFailWithError", params)
    }

    override fun onDetachedFromWindow() {
        debugLog("BBShortsView") { "onDetachedFromWindow - cleaning up shorts" }
        removeShorts()
        super.onDetachedFromWindow()
    }
}

File: ios/BBShortsView.swift
import BBNativePlayerKit
import bbnativeshared
import os

// MARK: - Logging Helper
private func log(_ message: String, level: LogLevel = .debug) {
  #if DEBUG
  // In debug builds, log everything
  NSLog("BBShorts: \(message)")
  #else
  // In release builds, only log warnings and errors
  if level == .warning || level == .error {
    NSLog("BBShorts: \(message)")
  }
  #endif
}

private enum LogLevel {
  case debug
  case info
  case warning
  case error
}

/**
 * React Native View for Blue Billywig Native Shorts
 *
 * This is a separate component from BBPlayerView specifically for Shorts playback.
 * Shorts use a different native SDK view (BBNativeShortsView) that supports
 * vertical swipe navigation and TikTok-style full-screen video experience.
 *
 * Usage:
 * ```tsx
 * <BBShortsView
 *   jsonUrl="https://demo.bbvms.com/sh/58.json"
 *   onDidSetupWithJsonUrl={(url) => console.log('Shorts loaded:', url)}
 *   onDidFailWithError={(error) => console.error('Error:', error)}
 * />
 * ```
 */
class BBShortsView: UIView, BBNativeShortsViewDelegate {
  private var shortsView: BBNativeShortsView?
  private var hasSetup: Bool = false

  // MARK: - Props (set from React Native)

  @objc var jsonUrl: String = "" {
    didSet {
      // If URL changed and we already have a shorts view, recreate it
      if oldValue != jsonUrl && hasSetup {
        log("jsonUrl changed from \(oldValue) to \(jsonUrl) - recreating shorts view")
        destroy()
        setupShortsIfNeeded()
      } else {
        setupShortsIfNeeded()
      }
    }
  }

  @objc var options: NSDictionary = [:] {
    didSet {
      setupShortsIfNeeded()
    }
  }

  // MARK: - Event handlers (RCTDirectEventBlock)

  @objc var onDidSetupWithJsonUrl: RCTDirectEventBlock?
  @objc var onDidFailWithError: RCTDirectEventBlock?
  @objc var onDidTriggerResize: RCTDirectEventBlock?

  // Override intrinsicContentSize to tell React Native this view wants to fill available space
  override var intrinsicContentSize: CGSize {
    return CGSize(width: UIView.noIntrinsicMetric, height: UIView.noIntrinsicMetric)
  }

  private func setupShortsIfNeeded() {
    // Only setup once we have both jsonUrl and we're in the window
    guard !jsonUrl.isEmpty, window != nil, !hasSetup else { return }
    hasSetup = true
    setupShorts()
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()

    if window != nil {
      log("BBShortsView.didMoveToWindow - view added to window")
      // Ensure this view respects its frame from React Native layout
      self.clipsToBounds = true

      // Find the parent view controller from the responder chain
      var parentVC: UIViewController?
      var responder = self.next
      while responder != nil {
        if let viewController = responder as? UIViewController {
          parentVC = viewController
          break
        }
        responder = responder?.next
      }

      if parentVC != nil {
        log("Found parent view controller: \(type(of: parentVC!))")
        // Try to setup if we have jsonUrl
        setupShortsIfNeeded()
      } else {
        log("WARNING - Could not find parent view controller!", level: .warning)
      }
    } else {
      log("BBShortsView.didMoveToWindow - view removed from window")
    }
  }

  private func setupShorts() {
    log("BBShortsView.setupShorts() called with jsonUrl: \(jsonUrl)")

    guard !jsonUrl.isEmpty else {
      log("Cannot setup shorts - jsonUrl is empty", level: .warning)
      return
    }

    // Find the parent view controller
    var parentVC: UIViewController?
    var responder = self.next
    while responder != nil {
      if let viewController = responder as? UIViewController {
        parentVC = viewController
        break
      }
      responder = responder?.next
    }

    guard let viewController = parentVC else {
      log("ERROR - Cannot setup shorts without parent view controller", level: .error)
      return
    }

    // Convert options to Swift dictionary
    var optionsDict: [String: Any]? = nil
    if let dict = options as? [String: Any], !dict.isEmpty {
      optionsDict = dict
    }

    // Create the shorts view using the factory method
    shortsView = BBNativeShorts.createShortsView(
      uiViewController: viewController,
      frame: bounds,
      jsonUrl: jsonUrl,
      options: optionsDict
    )

    guard let shorts = shortsView else {
      log("ERROR - Failed to create BBNativeShortsView", level: .error)
      onDidFailWithError?(["error": "Failed to create shorts view"])
      return
    }

    shorts.delegate = self

    // Add to view hierarchy with autolayout
    addSubview(shorts)
    shorts.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      shorts.topAnchor.constraint(equalTo: topAnchor),
      shorts.leadingAnchor.constraint(equalTo: leadingAnchor),
      shorts.trailingAnchor.constraint(equalTo: trailingAnchor),
      shorts.bottomAnchor.constraint(equalTo: bottomAnchor)
    ])

    log("BBShortsView.setupShorts() completed")
  }

  func destroy() {
    log("BBShortsView.destroy() called")
    shortsView?.destroy()
    shortsView?.removeFromSuperview()
    shortsView = nil
    hasSetup = false
  }

  deinit {
    destroy()
  }

  // MARK: - BBNativeShortsViewDelegate

  func bbNativeShortsView(shortsView: BBNativeShortsView, didSetupWithJsonUrl url: String?) {
    log("didSetupWithJsonUrl: \(url ?? "nil")")
    onDidSetupWithJsonUrl?(["url": url as Any])
  }

  func bbNativeShortsView(shortsView: BBNativeShortsView, didFailWithError error: String?) {
    log("didFailWithError: \(error ?? "nil")", level: .error)
    onDidFailWithError?(["error": error ?? "Unknown error"])
  }

  func bbNativeShortsView(shortsView: BBNativeShortsView, didTriggerResize width: Int, height: Int) {
    log("didTriggerResize: \(width)x\(height)")
    onDidTriggerResize?([
      "width": width,
      "height": height
    ])
  }
}
